File: src/feature-flags/feature-flag.service.ts
import { Inject, Injectable } from "@nestjs/common"
import { ConfigService } from "@nestjs/config"
import { Client, OpenFeatureClient } from "@openfeature/nestjs-sdk"
import type { DevCycleClient } from "@devcycle/nodejs-server-sdk"

@Injectable()
export class FeatureFlagService {
  private initialized = false
  private initializationPromise: Promise<void>

  constructor(
    private configService: ConfigService,
    @OpenFeatureClient() private client: Client,
    @Inject("DVC_CLIENT") private devcycleClient: DevCycleClient | null
  ) {
    this.initialized = !!this.devcycleClient
  }

  async isInitialized(): Promise<boolean> {
    return this.initialized
  }

  async waitForInitialization(): Promise<void> {
    // Already initialized in constructor
    return Promise.resolve()
  }

  async getBooleanValue(
    userId: string,
    key: string,
    defaultValue: boolean = false
  ) {
    if (!this.initialized) {
      return defaultValue
    }

    try {
      const context = {
        targetingKey: userId,
        user_id: userId,
      }

      const result = await this.client.getBooleanValue(
        key,
        defaultValue,
        context
      )

      return result
    } catch (error) {
      console.warn(`Error getting feature flag ${key}:`, error.message)
      return defaultValue
    }
  }

  async getStringValue(userId: string, key: string, defaultValue: string = "") {
    if (!this.initialized) {
      return defaultValue
    }

    try {
      const context = {
        targetingKey: userId,
        user_id: userId,
      }

      const result = await this.client.getStringValue(
        key,
        defaultValue,
        context
      )
      return result
    } catch (error) {
      console.warn(`Error getting string variable ${key}:`, error.message)
      return defaultValue
    }
  }

  async getNumberValue(userId: string, key: string, defaultValue: number = 0) {
    if (!this.initialized) {
      return defaultValue
    }

    try {
      const context = {
        targetingKey: userId,
        user_id: userId,
      }

      const result = await this.client.getNumberValue(
        key,
        defaultValue,
        context
      )
      return result
    } catch (error) {
      console.warn(`Error getting number variable ${key}:`, error.message)
      return defaultValue
    }
  }

  async getObjectValue(userId: string, key: string, defaultValue: any = {}) {
    if (!this.initialized) {
      return defaultValue
    }

    try {
      const context = {
        targetingKey: userId,
        user_id: userId,
      }

      const result = await this.client.getObjectValue(
        key,
        defaultValue,
        context
      )
      return result
    } catch (error) {
      console.warn(`Error getting object variable ${key}:`, error.message)
      return defaultValue
    }
  }

  // Legacy method for backward compatibility
  async getVariableValue(userId: string, key: string, defaultValue: any) {
    // Determine the type and call appropriate method
    if (typeof defaultValue === "boolean") {
      return this.getBooleanValue(userId, key, defaultValue)
    } else if (typeof defaultValue === "string") {
      return this.getStringValue(userId, key, defaultValue)
    } else if (typeof defaultValue === "number") {
      return this.getNumberValue(userId, key, defaultValue)
    } else {
      return this.getObjectValue(userId, key, defaultValue)
    }
  }

  // Additional method to access DevCycle client directly if needed
  getDevCycleClient(): DevCycleClient | null {
    return this.devcycleClient
  }

  // Method to get all features for a user (using DevCycle client directly)
  async getAllFeatures(userId: string) {
    if (!this.devcycleClient) {
      console.warn("DevCycle client not available")
      return {}
    }

    try {
      return await this.devcycleClient.allFeatures({ user_id: userId })
    } catch (error) {
      console.warn(`Error getting all features:`, error.message)
      return {}
    }
  }
}

File: java-backend/src/main/java/com/shopper/service/FeatureFlagService.java
package com.shopper.service;

import com.devcycle.sdk.server.common.model.DevCycleUser;
import com.devcycle.sdk.server.local.api.DevCycleLocalClient;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Service
@RequiredArgsConstructor
@Slf4j
public class FeatureFlagService {
    
    @Value("${app.devcycle.server-sdk-key}")
    private String devCycleServerSdkKey;
    
    private final DevCycleLocalClient devCycleLocalClient;
    private final Map<String, Object> fallbackFlags = new ConcurrentHashMap<>();
    private boolean initialized = false;
    
    private void ensureInitialized() {
        if (!initialized) {
            // Initialize with default feature flags for demo purposes (fallback)
            fallbackFlags.put("new-flow", false);
            fallbackFlags.put("premium-features", true);
            fallbackFlags.put("enhanced-product-details", true);
            fallbackFlags.put("beta-features", false);
            fallbackFlags.put("use-neon", false);
            
            initialized = true;
            log.info("Feature flags initialized with default values");
            
            if (devCycleLocalClient == null) {
                log.info("🔄 DevCycle client not available, using fallback values for all feature flags");
            }
        }
    }

    public boolean getBooleanValue(String userId, String key, boolean defaultValue) {
        ensureInitialized();
        
        if (devCycleLocalClient == null) {
            log.debug("🔄 DevCycle client not available, using fallback for '{}': {}", key, defaultValue);
            return defaultValue;
        }
        
        try {
            DevCycleUser user = DevCycleUser.builder().userId(userId).build();
            Boolean details = devCycleLocalClient.variableValue(user, key, defaultValue);
            
            boolean flagValue = details;
            log.debug("🎛️ Feature flag '{}' evaluated to: {} for user: {}", 
                     key, flagValue, userId);
            
            return flagValue;
            
        } catch (Exception e) {
            log.warn("Error getting feature flag '{}' for user '{}': {}, using default: {}", 
                    key, userId, e.getMessage(), defaultValue);
            return defaultValue;
        }
    }
    
    public String getStringValue(String userId, String key, String defaultValue) {
        ensureInitialized();
        
        if (devCycleLocalClient == null) {
            log.debug("🔄 DevCycle client not available, using fallback for '{}': {}", key, defaultValue);
            return defaultValue;
        }
        
        try {
            DevCycleUser user = DevCycleUser.builder().userId(userId).build();
            String flagValue = devCycleLocalClient.variableValue(user, key, defaultValue);
            
            log.debug("🎛️ Feature flag '{}' evaluated to: {} for user: {}", 
                        key, flagValue, userId);
            
            return flagValue;
            
        } catch (Exception e) {
            log.warn("Error getting feature flag '{}' for user '{}': {}, using default: {}", 
                    key, userId, e.getMessage(), defaultValue);
            return defaultValue;
        }
    }
    
    public Number getNumberValue(String userId, String key, Number defaultValue) {
        ensureInitialized();

        if (devCycleLocalClient == null) {
            log.debug("🔄 DevCycle client not available, using fallback for '{}': {}", key, defaultValue);
            return defaultValue;
        }
        
        try {
            DevCycleUser user = DevCycleUser.builder().userId(userId).build();
            Double flagValue = devCycleLocalClient.variableValue(user, key, defaultValue.doubleValue());
            
            log.debug("🎛️ Feature flag '{}' evaluated to: {} for user: {}", 
                     key, flagValue, userId);
            
            return flagValue;
            
        } catch (Exception e) {
            log.warn("Error getting feature flag '{}' for user '{}': {}, using default: {}", 
                    key, userId, e.getMessage(), defaultValue);
            return defaultValue;
        }
    }
    
    public Object getObjectValue(String userId, String key, Object defaultValue) {
        ensureInitialized();
        
        if (devCycleLocalClient == null) {
            log.debug("🔄 DevCycle client not available, using fallback for '{}': {}", key, defaultValue);
            return defaultValue;
        }
        
        try {
            DevCycleUser user = DevCycleUser.builder().userId(userId).build();
            Object flagValue = devCycleLocalClient.variableValue(user, key, defaultValue);
            
            log.debug("🎛️ Feature flag '{}' evaluated to: {} for user: {}", 
                     key, flagValue, userId);
            
            return flagValue;
            
        } catch (Exception e) {
            log.warn("Error getting feature flag '{}' for user '{}': {}, using default: {}", 
                    key, userId, e.getMessage(), defaultValue);
            return defaultValue;
        }
    }
    
    public Map<String, Object> getAllFeatures(String userId) {
        ensureInitialized();
        
        if (devCycleLocalClient == null) {
            log.debug("🔄 DevCycle client not available, returning fallback flags for user: {}", userId);
            return new HashMap<>(fallbackFlags);
        }
        
        try {
            Map<String, Object> features = new HashMap<>();
            
            DevCycleUser user = DevCycleUser.builder().userId(userId).build();
            
            features.put("new-flow", devCycleLocalClient.variableValue(user, "new-flow", false));
            features.put("premium-features", devCycleLocalClient.variableValue(user, "premium-features", true));
            features.put("enhanced-product-details", devCycleLocalClient.variableValue(user, "enhanced-product-details", true));
            features.put("beta-features", devCycleLocalClient.variableValue(user, "beta-features", false));
            features.put("use-neon", devCycleLocalClient.variableValue(user, "use-neon", false));
            
            log.debug("All features for user '{}': {}", userId, features);
            return features;
            
        } catch (Exception e) {
            log.warn("Error getting all features for user '{}': {}, using fallbacks", userId, e.getMessage());
            return new HashMap<>(fallbackFlags);
        }
    }
    
    public boolean isInitialized() {
        return initialized && (devCycleLocalClient == null || devCycleLocalClient.isInitialized());
    }
    
}